x-pack/platform/plugins/shared/onechat/server/routes/agent_profiles.ts

```diff
@@ -0,0 +1,235 @@
+          return response.notFound();
+        }
+        const { agents } = getInternalServices();
+        const service = await agents.getProfileService(request);
+
+        const profile = await service.get(request.params.id);
+        return response.ok({ body: profile });
+      })
+    );
+
+  // Create agent profile
+  router.versioned
+    .post({
+      path: '/api/chat/agents/profiles',
+      security: {
+        authz: { requiredPrivileges: [apiPrivileges.manageOnechat] },
+      },
+      access: 'public',
+      summary: 'Create an agent profile',
+      description: TECHNICAL_PREVIEW_WARNING,
+      options: {
+        tags: ['agent'],
+        availability: {
+          stability: 'experimental',
+        },
+      },
+    })
+    .addVersion(
+      {
+        version: '2023-10-31',
+        validate: {
+          request: {
+            body: schema.object({
+              id: schema.string(),
+              name: schema.string(),
+              description: schema.string(),
+              customInstructions: schema.string(),
+              toolSelection: schema.arrayOf(
+                schema.object({
+                  provider: schema.maybe(schema.string()),
+                  toolIds: schema.arrayOf(schema.string()),
+                })
+              ),
+            }),
+          },
+        },
+      },
+      wrapHandler(async (ctx, request, response) => {
+        const { uiSettings } = await ctx.core;
+        const enabled = await uiSettings.client.get(ONECHAT_AGENT_API_UI_SETTING_ID);
+        if (!enabled) {
+          return response.notFound();
+        }
+        const { agents } = getInternalServices();
+        const service = await agents.getProfileService(request);
+        const profile = await service.create(request.body);
+        return response.ok({ body: profile });
+      })
+    );
+
+  // Update agent profile
+  router.versioned
+    .put({
+      path: '/api/chat/agents/profiles/{id}',
+      security: {
+        authz: { requiredPrivileges: [apiPrivileges.manageOnechat] },
+      },
+      access: 'public',
+      summary: 'Update an agent profile',
+      description: TECHNICAL_PREVIEW_WARNING,
+      options: {
+        tags: ['agent'],
+        availability: {
+          stability: 'experimental',
+        },
+      },
+    })
+    .addVersion(
+      {
+        version: '2023-10-31',
+        validate: {
+          request: {
+            params: schema.object({ id: schema.string() }),
+            body: schema.object({
+              name: schema.maybe(schema.string()),
+              description: schema.maybe(schema.string()),
+              customInstructions: schema.maybe(schema.string()),
+              toolSelection: schema.maybe(
+                schema.arrayOf(
+                  schema.object({
+                    provider: schema.maybe(schema.string()),
+                    toolIds: schema.arrayOf(schema.string()),
+                  })
+                )
+              ),
```

Comment:
NIT: factorize schema for `toolSelection`, reduces risk of error if we introduce addition sub types for `ToolSelection`